kernel/serial_port/serial_port_arm/src/lib.rs

```diff
@@ -0,0 +1,51 @@
+#![no_std]
+
+#[macro_use]
+extern crate cfg_if;
+
+cfg_if ! {
+    if #[cfg(target_vendor = "stm32f407")] {
+        extern crate stm32f4_discovery;
+        pub use stm32f4_discovery::uart::{get_serial_port, SerialPort, SerialPortAddress};
+    } 
+
+    // Dummy implementation for when no physical device is present, in which case semihosting will be used
```

Comment:
Please add a block of crate-level documentation explaining what semihosting is and why the complex target_vendor-specific configuration is required here.